File: Python/multi_datalogger_reader.py
"""

 For running with Joe's Simulator:
    sudo python multi_datalogger_reader.py --port 1045 --ip 192.168.7.2

 For running in HARP lab:
    sudo python multi_datalogger_reader.py

 testing program for getting data from 4 ch HARP 3B04 230307
 two channels at 200kHz/ch


 In order for the changes that this function makes be observable across both threads, this function relies on global variables

"""

import struct
import socket
import threading
import queue  
import numpy as np
import sys
import argparse
import time
import datetime
import psutil
import os
import traceback
from scipy.signal import filtfilt, ellip, freqz, lfilter
from process_data import  ThresholdDetect, SegmentPulses, PreprocessSegmentInterleaved, PreprocessSegmentStacked, SaveDataSegment, WritePulseAmplitudes
from TDOA_estimation import EllipticFilter, GCC_PHAT, CrossCorr, GenerateFIR_Filter
from utils import SetHighPriority

print('This code has been tested for python version 3.11.6, your version is:', sys.version)

parser = argparse.ArgumentParser(description='Program command line arguments')
parser.add_argument('--port', default=50000, type=int)                           # port to listen to for UDP packets
parser.add_argument('--ip', default="192.168.100.220",type=str)                  # IP address of data logger (simulator)
parser.add_argument('--fw', default = 1550, type=int)                            # firmware version
parser.add_argument('--output_file', default = "clicks_data.txt", type=str)      # output file for logging time and peak amp. of pulses
args = parser.parse_args()
output_file = open(args.output_file, 'w')                 # clear contents in output file

UDP_IP = args.ip                                          # IP address of data logger or simulator 
UDP_PORT = args.port                                      # Port to listen for UDP packets

print('Listening to IP address, ', UDP_IP,' and port ',UDP_PORT)

SetHighPriority()            # set the program to run at a high priority on the system (nice value = -15)

### import variables according to firmware version specified ###
print('Assuming firmware version: ', args.fw)
if args.fw == 1550:
    from Firmware_config.firmware_1550 import *
    PreprocessSegment = PreprocessSegmentInterleaved
elif args.fw == 1240:
    from Firmware_config.firmware_1240 import *
    PreprocessSegment = PreprocessSegmentInterleaved
else:
    print('ERROR: Unknown firmware version')
    sys.exit()  # Exiting the program

TIME_WINDOW = .01                                                         # fraction of a second to consider  
NUM_PACKS_DETECT = round(TIME_WINDOW * SAMPLE_RATE / SAMPS_PER_CHANNEL)  # the number of data packets that are needed to perform energy detection 

print('Bytes per packet:       ',         REQUIRED_BYTES)
print('Microseconds between packets:   ', MICRO_INCR)
print('Number of channels:     ',         NUM_CHAN)
print('Data bytes per channel: ',         DATA_BYTES_PER_CHANNEL)
print("Detecting over a time window of ", TIME_WINDOW," seconds, using ",NUM_PACKS_DETECT, " packets") 

def restartListener():
    """
    Functionality:
    This function is responsible for (re)starting the listener.
    The socket connection is re(set). The  buffer (dataBuffer) and segment to be processed (dataSegment) are cleared.

    """
    
    global dataBuffer, dataSegment, dataTimes, udpSocket
    
    with udpSocketLock:
        udpSocket.close()
        udpSocket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        if UDP_IP == "192.168.100.220": # IP address of the data logger
            print('Sending wake up data to IP address: ', UDP_IP)

            # need 100 bytes to get Open command through
            m1 = b'Open'
            m2 = bytearray(np.zeros((1,96),dtype=int))
            udpSocket.sendto(m1 + m2, (UDP_IP, UDP_PORT))
        elif UDP_IP == "192.168.7.2": # IP address of Joe's simulator
            udpSocket.bind((UDP_IP, UDP_PORT))
            print('Listening to simulator')
        elif UDP_IP == "192.168.7.7": # IP address of Joe's simulator
            udpSocket.bind((UDP_IP, UDP_PORT))
        elif UDP_IP == "self": # IP address of Joe's simulator
            udpSocket.bind(("127.0.0.1", UDP_PORT))
        else:
            print("ERROR: Unknown IP address" )
            sys.exit()
    
    with dataBufferLock:
        dataBuffer = queue.Queue()
    with dataSegmentLock:
        dataSegment = np.array([])
    with dataTimesLock:
        dataTimes = np.array([])

def UdpListener():
    """
    Functionality:
    This function serves as a UDP listener that continuously listens for incoming UDP packets.
    It receives UDP packets of a specified size, checks their length, and if they meet the expected size, 
    increments a packet counter. The received data is then placed into a shared buffer for further processing.

    Global Variables:
    - dataBuffer: shared buffer to store received data packets.
    - packetCounter: counter to keep track of the number of received packets.
    - udpSocket: the UDP socket used for listening to incoming packets.    
    """
    try:
        global dataBuffer, packetCounter, udpSocket
        startPacketTime = time.time()
        printInterval = 500
        while not errorEvent.is_set():
            
            
            with udpSocketLock:
                dataBytes, addr1 = udpSocket.recvfrom(PACKET_SIZE + 1)  # + 1 to detect if more bytes are received
                #print(dataBytes)
                #break
            packetCounter += 1
            if packetCounter % printInterval == 0:
                with dataBufferLock:
                    qSize = dataBuffer.qsize()
                print("Num packets received is ", packetCounter, (time.time() - startPacketTime)/printInterval,qSize,packetCounter - qSize)
                startPacketTime = time.time()
            with dataBufferLock:
                dataBuffer.put(dataBytes)  # Put received data into the buffer
    except Exception as e:
        print(f"Error occurred in UDP listener thread: {e}")
        traceback.print_exc()  # Print detailed traceback information
        errorEvent.set()

def DataProcessor():
    """
    Functionality:
    This function serves as a data processor that continuously processes data segments retrieved from a shared buffer (dataBuffer).
    It extracts the necessary information from the received data, such as timestamps and sample values, performs adjustments,
    and stores the processed data into a segment (dataSegment). This segment is then processed.

    Global Variables:
    - dataBuffer: Shared buffer containing received data packets.
    - dataSegment: Segment of data to be processed.
    - dataTimes: Array containing timestamps associated with data segments. 
    """
    try:
        global dataBuffer, dataSegment, dataTimes
        
        ### define IIR filter - should be passed as function arguement
        '''
        order = 4
        ripple_dB = 0.1
        cutoffFrequency = 10000
        b, a = EllipticFilter(order, ripple_dB, cutoffFrequency, SAMPLE_RATE)
        print('b: ', b)
        print('a: ', a)
        '''

        ### define FIR filter - should be passed as function arguement 
        cutoffFrequency = 10000  # Cutoff frequency for highpass filter
        numTaps = 31  # Number of taps for the FIR filter
        taps = GenerateFIR_Filter(cutoffFrequency, numTaps, SAMPLE_RATE)
        print(taps)


        previousTime = False # initalize the previous packet time to False, update with every new packet
        while not errorEvent.is_set():

            # begin new data segment 
            with dataSegmentLock:
                dataSegment = np.array([])
            with dataTimesLock:
                dataTimes = np.array([])
            
            while len(dataSegment) < (NUM_PACKS_DETECT * SAMPS_PER_CHANNEL * NUM_CHAN):
                
                with dataBufferLock:
                    qSize = dataBuffer.qsize()
                if qSize < 1:
                    print("sleeping")
                    time.sleep(.2)
                    continue
                
                dataBytes = dataBuffer.get()
                
                ### check packet length
                if len(dataBytes) != PACKET_SIZE:
                    print('Error: recieved incorrect number of packets')
                    previousTime = False
                    restartListener()
                    continue

                ####
                # Unpacks the binary dataBytes into a tuple according to a specified format,
                # where 'B' denotes unsigned char (1 byte) and 'H' denotes unsigned short (2 bytes),
                # with the format dynamically constructed based on the length of dataBytes and HEAD_SIZE.
                # '>' is a format character that specifies big-endian byte order.
                dataUnpacked = struct.unpack('>' + 'B'*HEAD_SIZE + 'H'*((len(dataBytes) - HEAD_SIZE) // 2), dataBytes)
                ####

                dataTime = dataUnpacked[:HEAD_SIZE]
                dataSamples = np.array(dataUnpacked[HEAD_SIZE:]) - 2**15  # Extract and adjust dataSamples
                #print("dataSamples: ",dataSamples)
                #return
                us = (dataBytes[6],dataBytes[7],dataBytes[8],dataBytes[9])
                microSeconds = int.from_bytes(us,'big')         # get micro seconds from dataTime (unsigned char ist)
                dateTime = datetime.datetime(2000+dataTime[0], dataTime[1], dataTime[2], dataTime[3], dataTime[4], dataTime[5], microSeconds, tzinfo=datetime.timezone.utc)
                
                ### compare the current packet's time to the previous packet time
                if previousTime and ((dateTime - previousTime).microseconds != MICRO_INCR):
                    print("Error: Time not incremented by ", MICRO_INCR)
                    previousTime = False
                    restartListener()
                    continue

                with dataTimesLock:
                    dataTimes = np.append(dataTimes, dateTime) 
                with dataSegmentLock:
                    dataSegment = np.append(dataSegment,dataSamples)

                previousTime = dateTime
            
            #if np.random.random() < .001:
            #    adsfasdfsaf

            with dataSegmentLock:
                ch1, ch2, ch3, ch4 = PreprocessSegment(dataSegment, NUM_PACKS_DETECT, NUM_CHAN, SAMPS_PER_CHANNEL)
            with dataTimesLock:
                #values = SegmentPulses(ch1, dataTimes, SAMPLE_RATE, 2500, False) # Set true to save segmented pulses
                values = ThresholdDetect(ch1,dataTimes, SAMPLE_RATE, 80)
            if values == None: # if no pulses were detected to segment, then get next segment
                continue
           
            clickTimes, clickAmplitudes, clickStartPoints, clickEndPoints = values
            #SaveDataSegment(clickTimes[0], dataSegment, ch1, ch2, ch3, ch4) 
            #print("pulse detected! ", clickAmplitudes[0])
            WritePulseAmplitudes(clickTimes, clickAmplitudes, args.output_file)
            
            ### IIR code
            '''
            ch1Filtered = filtfilt(b, a, ch1)
            ch2Filtered = filtfilt(b, a, ch2)
            ch3Filtered = filtfilt(b, a, ch3)
            ch4Filtered = filtfilt(b, a, ch4)
            '''

            ### FIR code
            #print('ch1 examples: ', ch1[:10])
            prefiltTime = time.time()            
            ch1 = lfilter(taps, 1.0, ch1)
            ch2 = lfilter(taps, 1.0, ch2)
            ch3 = lfilter(taps, 1.0, ch3)
            ch4 = lfilter(taps, 1.0, ch4)
            #print('P FIR: ', time.time() - prefiltTime)
            #print('ch1 filt examples: ', ch1[:10])

            dataMatrixFiltered = np.vstack(np.array([ch1, ch2, ch3, ch4]))
            
            gccStart = time.time()
            tdoaEstimates = GCC_PHAT(dataMatrixFiltered, SAMPLE_RATE, max_tau=None, interp=1)
            print("P GCC: ", time.time() - gccStart)
            #print('here')
            #print(tdoaEstimates)
            
    except Exception as e:
        print(f"Error occurred in Data Processor thread: {e}")
        traceback.print_exc()  # Print detailed traceback information
        errorEvent.set()

### In order for the changes that one thread makes to shared variables be observable across both threads, global variables are needed

# Global Variables: counter and socket 
packetCounter = 0                                                  # counter for the number of packets received
udpSocket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)       # udp socket that uses IPV4

# Global variables: data containers
dataBuffer = queue.Queue()                                         # buffer (Queue) for communication between threads
dataSegment = np.array([])                                         # decoded data pulled from buffer to be processed
dataTimes = np.array([])                                           # timestamps associated with data inside dataSegment

# Global Variables: locks
dataBufferLock = threading.Lock()                                  # a lock for safe thread access to variable "dataBuffer"
dataSegmentLock = threading.Lock()                                 # a lock for safe thread access to variable "dataSegment"
dataTimesLock = threading.Lock()                                   # a lock for safe thread access to variable "dataTimes"
udpSocketLock = threading.Lock()                                   # a lock for safe thread access to variable "udpSocket"

errorEvent = threading.Event()

while True: # Loop continuously so when a thread errors (errorEvent.set() occurs) threads will close and restart 
 
    restartListener()
    udpThread = threading.Thread(target=UdpListener)               # create a thread for listening for incoming UDP packets
    processorThread = threading.Thread(target=DataProcessor)       # create a thread for processing UDP packet data

    udpThread.start()
    processorThread.start()
    
    udpThread.join()
    processorThread.join()

    print("Restarting threads...")

    errorEvent = threading.Event()
    
    # Global variables: data containers
    dataBuffer = queue.Queue()                                     # buffer (Queue) for communication between threads
    dataSegment = np.array([])                                     # decoded data pulled from buffer to be processed
    dataTimes = np.array([])                                       # timestamps associated with data inside dataSegment

File: C/multi_datalogger_reader.cpp
/*
This is a C++ version of Python/multi_datalogger_reader.py

Compile code manually:
g++ multi_datalogger_reader.cpp process_data.cpp utils.cpp -o listen -larmadillo
g++ -std=c++20 multi_datalogger_reader.cpp process_data.cpp utils.cpp -o listen -larmadillo -I/usr/include/sigpack-1.2.7/sigpack
g++ -std=c++20 multi_datalogger_reader.cpp process_data.cpp utils.cpp TDOA_estimation.cpp -o listen -larmadillo -I/usr/include/sigpack-1.2.7/sigpack -lliquid
g++ -Ofast -std=c++20 -march=native -flto multi_datalogger_reader.cpp process_data.cpp utils.cpp TDOA_estimation.cpp -o listen -DARMA_DONT_USE_WRAPPER -DNDEBUG -lopenblas -llapack -I/usr/include/sigpack-1.2.7/sigpack -lliquid
g++ -Ofast -std=c++20 -march=native -DEIGEN_FFT_BACKEND=FFTW -flto -msse2 multi_datalogger_reader.cpp process_data.cpp utils.cpp TDOA_estimation.cpp -o listen -DARMA_DONT_USE_WRAPPER -DNDEBUG -lopenblas -llapack -I/usr/include/sigpack-1.2.7/sigpack -lliquid -lfftw3
g++ -Ofast -std=c++20 -march=native -DEIGEN_FFT_BACKEND=FFTW -flto -msse2 multi_datalogger_reader.cpp process_data.cpp utils.cpp TDOA_estimation.cpp -o listen -lopenblas -llapack -I/usr/include/sigpack-1.2.7/sigpack -lliquid -lfftw3
g++ -Ofast -std=c++20 -march=native -flto multi_datalogger_reader.cpp process_data.cpp utils.cpp TDOA_estimation.cpp -o listen -lopenblas -llapack -I/usr/include/sigpack-1.2.7/sigpack -lliquid -lfftw3

Execute (datalogger simulator):
./listen 192.168.7.2 1045 1240

Execute (datalogger):
./listen.exe 192.168.100.220 50000 1240

TO DO:

RESOURCES:
    debugging (core dump): https://www.youtube.com/watch?v=3T3ZDquDDVg&t=190s
*/

#include <iostream>
#include <cstring>
#include <cstdio>
#include <ostream>
#include <string>
#include <fstream>
#include <mutex>
#include <thread>
#include <queue>
#include <vector>
#include <chrono>
#include <atomic>
#include <stdexcept>
#include <random>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <unistd.h>
#include <iomanip> //put_time
#include <ctime>
#include <cstdint>
#include <armadillo>

#include <sigpack.h>
#include <fftw/fftw.h>
//#include <Eigen/Dense>
#include <eigen3/Eigen/Dense>
#include <eigen3/Eigen/Core>

#include <complex>
#include <liquid/liquid.h>
//#include <liquid.h>

#include "process_data.h"
#include "TDOA_estimation.h"
#include "utils.h"
#include "my_globals.h"
#include "filters.h"

using std::cout;
using std::cin;
using std::endl;
using std::string;
using std::cerr;
using std::vector;
using std::ifstream;
using namespace std::chrono_literals;

// Global Constants
unsigned int HEAD_SIZE;                                 //packet head size (bytes)
unsigned int NUM_CHAN;                                  //number of channels per packet
unsigned int SAMPS_PER_CHANNEL;                         //samples per packet per channel, for 2 channels, this value is 5*62  = 310
unsigned int BYTES_PER_SAMP;                            //bytes per sample
unsigned int DATA_SIZE;                                 //packet data size (bytes)
unsigned int PACKET_SIZE;                               //packet size (bytes)
unsigned int REQUIRED_BYTES;
unsigned int DATA_BYTES_PER_CHANNEL;                    //number of data bytes per channel (REQUIRED_BYTES - 12) / 4 channels
unsigned int NUM_PACKS_DETECT;
unsigned int DATA_SEGMENT_LENGTH;
unsigned int MICRO_INCR;                                // time between packets
const unsigned int SAMPLE_RATE = 1e5;
const double TIME_WINDOW = 0.01;                 // fraction of a second to consider  
const string OUTPUT_FILE = "clicks_data.txt";
std::string outputFile = "Ccode_clicks.txt"; // Change to your desired file name

int packetCounter = 0;

void UdpListener(Session& sess) {
    /*
    Functionality:
    This function serves as a UDP listener that continuously listens for incoming UDP packets.
    It receives UDP packets of a specified size, checks their length, and if they meet the expected size, 
    increments a packet counter. The received data is then placed into a shared buffer for further processing.

    Global Variables:
    - dataBuffer: shared buffer to store received data packets.
    - packetCounter: counter to keep track of the number of received packets.
    - udpSocket: the UDP socket used for listening to incoming packets.    
    */

    try {
        struct sockaddr_in addr;
        socklen_t addrLength = sizeof(addr);
        int bytesReceived;
        int printInterval = 500;
        int receiveSize = PACKET_SIZE + 1;      // + 1 to detect if an erroneous amount of data is being sent
        vector<uint8_t> dataBytes(receiveSize);
        std::chrono::duration<double> durationPacketTime;
        auto startPacketTime = std::chrono::steady_clock::now();
        auto endPacketTime = startPacketTime;
        int qSize;
        double define;
        while (!sess.errorOccurred) {
            
            sess.udpSocketLock.lock();
            bytesReceived = recvfrom(sess.datagramSocket, dataBytes.data(), receiveSize, 0, (struct sockaddr*)&addr, &addrLength);
            sess.udpSocketLock.unlock();
            
            if (bytesReceived == -1) {
                cerr << "Error in recvfrom" << endl;
                continue;
            }
            dataBytes.resize(bytesReceived); // Adjust size based on actual bytes received
            packetCounter += 1;
            if (packetCounter % printInterval == 0) {
                endPacketTime = std::chrono::steady_clock::now();
                durationPacketTime = endPacketTime - startPacketTime;
                sess.dataBufferLock.lock();
                qSize = sess.dataBuffer.size();
                sess.dataBufferLock.unlock();
                define = durationPacketTime.count() / printInterval; 
                cout << "Num packets received is " <<  packetCounter << " " << define  << " " << qSize << " " << packetCounter - qSize << endl;
                startPacketTime = std::chrono::steady_clock::now();
            }
            sess.dataBufferLock.lock();
            sess.dataBuffer.push(dataBytes);
            sess.dataBufferLock.unlock();
        }
    } catch (const std::exception& e ){
        // Handle the exception
        cerr << "Error occured in UDP Listener Thread: " << endl;
        cerr << e.what() << endl;
        sess.errorOccurred = true;
    }
}

void DataProcessor(void (*ProcessingFunction)(vector<double>&, vector<TimePoint>&, const string&, 
arma::Col<double>&, arma::Col<double>&, arma::Col<double>&, arma::Col<double>&), Session& sess) {
    /*
    Functionality:
    This function serves as a data processor that continuously processes data segments retrieved from a shared buffer (dataBuffer).
    It extracts the necessary information from the received data, such as timestamps and sample values, performs adjustments,
    and stores the processed data into a segment (dataSegment). This segment is then processed.

    Global Variables:
    - dataBuffer: Shared buffer containing received data packets.
    - dataSegment: Segment of data to be processed.
    - dataTimes: Array containing timestamps associated with data segments. 
    */

    try {
        
        int channelSize = DATA_SEGMENT_LENGTH / NUM_CHAN; // the number of samples per channel within a dataSegment
        
        // declare FFT object
        sp::FFTW fftw(channelSize, FFTW_ESTIMATE); // no 0 padding is currently being used
        
        // Define FIR filter using the liquid library
        arma::Col<double> h = { 8.5304705e-18, -1.2040846e-03, -2.7904883e-03, -4.2366693e-03,
         -3.9514871e-03, -9.6724173e-18,  8.2750460e-03,  1.8624326e-02,
          2.5445996e-02,  2.1282293e-02,  2.4025036e-17, -3.9675705e-02,
         -9.2092186e-02, -1.4542012e-01, -1.8531199e-01,  8.0040973e-01,
         -1.8531199e-01, -1.4542012e-01, -9.2092186e-02, -3.9675705e-02,
          2.4025036e-17,  2.1282293e-02,  2.5445996e-02,  1.8624326e-02,
          8.2750460e-03, -9.6724173e-18, -3.9514871e-03, -4.2366693e-03,
         -2.7904883e-03, -1.2040846e-03,  8.5304705e-18};         // filter coefficients
        
        // Define FIR filter using the liquid library 
        //firfilt_rrrf q = firfilt_rrrf_create(h,h.n_elem);// create filter object
        
        
        // Define IIR filter using SigPack library
        /*
        sp::IIR_filt<double, double, double> iir_filt;
        arma::Col<double> b_iir = {0.49580191, -1.9260157, 2.8613285, -1.9260157, 0.49580191};
        arma::Col<double> a_iir = {1.0, -2.53934052, 2.67821627, -1.31270499, 0.26392123};
        iir_filt.set_coeffs(b_iir,a_iir);
        */

        // Define FIR filter using SigPack library
        sp::FIR_filt<double, double, double> fir_filt;
        fir_filt.set_coeffs(h);

        bool previousTimeSet = false;
        std::chrono::time_point<std::chrono::system_clock> previousTime = std::chrono::time_point<std::chrono::system_clock>::min(); // CHECK VALUE

        // Initialize armadillo containers to be used for number crunching
        arma::Col<double> ch1(channelSize);
        arma::Col<double> ch2(channelSize);
        arma::Col<double> ch3(channelSize);
        arma::Col<double> ch4(channelSize);
        arma::Mat<double> dataMatrix(ch1.n_elem, 4);
        
        vector<uint8_t> dataBytes;
        while (!sess.errorOccurred) {
            
            sess.dataTimesLock.lock();
            sess.dataTimes.clear();
            sess.dataTimesLock.unlock();
            
            sess.dataSegmentLock.lock();
            sess.dataSegment.clear();
            sess.dataSegmentLock.unlock();

            //arma::Col<double> pad(15, arma::fill::zeros);

            while (sess.dataSegment.size() < DATA_SEGMENT_LENGTH) {
                
                sess.dataBufferLock.lock();
                int qSize = sess.dataBuffer.size();
                sess.dataBufferLock.unlock();
                //cout << "qSize before: " << qSize << endl;
                if (qSize < 1){
                    cout << "Sleeping: " << endl;
                    std::this_thread::sleep_for(200ms);
                    continue;
                }
                
                sess.dataBufferLock.lock();
                dataBytes = sess.dataBuffer.front();
                sess.dataBuffer.pop();
                sess.dataBufferLock.unlock();

                if (dataBytes.size() != PACKET_SIZE){
                    cerr << "Error: recieved incorrect number of packets" << endl;
                    previousTimeSet = false; 
                    restartListener(sess);
                    continue;
                }
                
                std::tm timeStruct{};
                timeStruct.tm_year = (int)dataBytes[0] + 2000 - 1900;     // Offset for year since 2000.. tm_year is years since 1900 
                timeStruct.tm_mon = (int)dataBytes[1] - 1;                     // Months are 0-indexed
                timeStruct.tm_mday = (int)dataBytes[2];
                timeStruct.tm_hour = (int)dataBytes[3];
                timeStruct.tm_min = (int)dataBytes[4];
                timeStruct.tm_sec = (int)dataBytes[5];


                int64_t microSec = (static_cast<int64_t>(dataBytes[6]) << 24) +
                         (static_cast<int64_t>(dataBytes[7]) << 16) +
                         (static_cast<int64_t>(dataBytes[8]) << 8) +
                         static_cast<int64_t>(dataBytes[9]);
                                
                
                // Use std::mktime to convert std::tm to std::time_t
                std::time_t timeResult = std::mktime(&timeStruct);

                if (timeResult == std::time_t(-1)) {
                    cerr << "failure in mktime!! " << endl;
                }


                std::chrono::time_point<std::chrono::system_clock> specificTime = std::chrono::system_clock::from_time_t(timeResult);
                specificTime += std::chrono::microseconds(microSec);
                
                
                auto duration = specificTime - previousTime;
                auto elapsed_time_ms = std::chrono::duration_cast<std::chrono::microseconds>(specificTime - previousTime).count();

                if (previousTimeSet && (elapsed_time_ms != MICRO_INCR)){
                    cerr << "Error: Time not incremented by " <<  MICRO_INCR << " " << elapsed_time_ms <<  endl; 
                    previousTimeSet = false;
                    restartListener(sess);
                    continue;
                }
                
                // Convert byte data to doubles
                sess.dataSegmentLock.lock();
                ConvertData(sess.dataSegment, dataBytes, DATA_SIZE, HEAD_SIZE);
                sess.dataSegmentLock.unlock();
                return;
                //cout << "size: " << sess.dataSegment.size() << endl;
                /*cout << "chan 1 examples:  ";
                for (int yy = 0; yy < 10; yy++){
                    cout << ch1(yy) << " "; 
                }
                cout << endl;
                */
                sess.dataTimesLock.lock();
                sess.dataTimes.push_back(specificTime);
                sess.dataTimesLock.unlock();
                
                previousTime = specificTime;
                previousTimeSet = true;
                
                /*
                if (withProbability(0.001)){
                    //throw std::runtime_error("An error occurred");
                    sess.dataSegmentLock.lock();
                    sess.dataSegment[2] =  std::numeric_limits<double>::quiet_NaN();
                    sess.dataSegmentLock.unlock();
                }
                */
                
                

            }
            
            sess.dataSegmentLock.lock();
            ProcessingFunction(sess.dataSegment, sess.dataTimes, OUTPUT_FILE, ch1, ch2, ch3, ch4);
            sess.dataSegmentLock.unlock();
            
            double threshold = 80.0;
            
            DetectionResult values = ThresholdDetect(ch1, sess.dataTimes, threshold);
            if (values.maxPeakIndex < 0){
                continue;
            }
            WritePulseAmplitudes(values.peakAmplitude, values.peakTimes, outputFile);
            
            
            auto beforeFilter = std::chrono::steady_clock::now();
            FilterWithFIR(ch1,ch2,ch3,ch4, fir_filt);
            //filterWithLiquidFIR(ch1,ch2,ch3,ch4, fir_filt);
            //filterWithIIR(ch1,ch2,ch3,ch4, iir_filt);
            auto afterFilter = std::chrono::steady_clock::now();
            std::chrono::duration<double> durationFilter = afterFilter - beforeFilter;
            //cout << "C FIR Filter: " << durationFilter.count() << endl;
            


            /*cout << "chan 1 filt  examples:  ";
            for (int y = 0; y < 10; y++){
                cout << ch1(y) << " "; 
            }
            cout << endl;
            */
            
            dataMatrix.insert_cols(0,ch1);
            dataMatrix.insert_cols(1,ch2);
            dataMatrix.insert_cols(2,ch3);
            dataMatrix.insert_cols(3,ch4);
            
            int interp = 1;
            auto beforeGCC = std::chrono::steady_clock::now();
            arma::Mat<double> resultMatrix = GCC_PHAT(dataMatrix, interp, fftw, channelSize);
            //Eigen::MatrixXd resultMatrix = GCC_PHAT_Eigen(dataE, interp); // need to create dataE matrix 
            auto afterGCC = std::chrono::steady_clock::now();
            std::chrono::duration<double> durationGCC = afterGCC - beforeGCC;
            //cout << "C GCC: " << durationGCC.count() << endl;
            

            // Print the matrix (optional)
            for (int ii = 0; ii < 4; ++ii) {
                for (int jj = 0; jj < 4; ++jj) {
                    cout << resultMatrix(ii, jj) << " ";
                }
                cout << endl;
            }
        }
    } catch (const std::exception& e ){
        // Handle the exception
        cerr << "Error occured in data processor thread" << endl;
        cerr << e.what() << endl;
        sess.errorOccurred = true;
      }
}


int main(int argc, char *argv[]){
    arma::arma_version ver;
    cout << "ARMA version: "<< ver.as_string() << endl;

    string UDP_IP = argv[1];                                         // IP address of data logger or simulator
    if (UDP_IP == "self"){
        UDP_IP = "127.0.0.1";
    }
    int UDP_PORT = std::stoi(argv[2]);                                     // Port to listen for UDP packets
    int firmwareVersion = std::stoi(argv[3]);
    
    printf("Listening to IP address %s and port %d \n", UDP_IP.c_str(),UDP_PORT);

    //import variables according to firmware version specified
    cout << "Assuming firmware version: " << firmwareVersion << endl;
    void(*ProcessFncPtr)(vector<double>&, vector<TimePoint>&, const string&, arma::Col<double>&, arma::Col<double>&, arma::Col<double>&, arma::Col<double>&) = nullptr;
    if (firmwareVersion == 1550){
        ProcessFile("ConfigFiles/1550_config.txt");
        ProcessFncPtr = ProcessSegmentInterleaved;
    }
    else if (firmwareVersion == 1240){
        ProcessFile("ConfigFiles/1240_config.txt");
        ProcessFncPtr = ProcessSegmentInterleaved;
    }
    else{
        cerr << "ERROR: Unknown firmware version" << endl;
        return 1;
    }
    
    NUM_PACKS_DETECT = (int)(TIME_WINDOW * 100000 / SAMPS_PER_CHANNEL);  // NEED TO ROUND THIS  the number of data packets that are needed to perform energy detection 
    DATA_SEGMENT_LENGTH = NUM_PACKS_DETECT * SAMPS_PER_CHANNEL * NUM_CHAN; 

    cout << "HEAD_SIZE: "              << HEAD_SIZE               << endl; 
    cout << "SAMPS_PER_CHAN: "         << SAMPS_PER_CHANNEL       << endl;
    cout << "BYTES_PER_SAMP: "         << BYTES_PER_SAMP          << endl;
    cout << "Bytes per packet:       " << REQUIRED_BYTES         << endl;
    cout << "Time between packets:   " << MICRO_INCR             << endl;
    cout << "Number of channels:     " << NUM_CHAN               << endl;
    cout << "Data bytes per channel: " << DATA_BYTES_PER_CHANNEL << endl;
    cout << "Detecting over a time window of " << TIME_WINDOW << " seconds, using " << NUM_PACKS_DETECT <<  " packets" << endl;

    
    // Open the file in write mode and clear its contents if it exists, create a new file otherwise
    std::ofstream file(outputFile, std::ofstream::out | std::ofstream::trunc);
    if (file.is_open()) {
        file << "Timestamp (microseconds)" << std::setw(20) << "Peak Amplitude" << endl;
        file.close();
        cout << "File created and cleared: " << outputFile << endl;
    } 
    else {
        cerr << "Error: Unable to open file for writing: " << outputFile << endl;
        return 1;
    }

    
    while (true){
        Session sess;
        sess.UDP_PORT = UDP_PORT;
        sess.UDP_IP = UDP_IP;
        
        restartListener(sess);
        
        std::thread udpThread(UdpListener, std::ref(sess));
        std::thread processorThread(DataProcessor, ProcessFncPtr, std::ref(sess));

        udpThread.join();
        processorThread.join();
       
        if (sess.errorOccurred){
            cout << "Restarting threads..." << endl;
        }
        else {
            cout << "Unknown problem occurred" << endl;
        }
    }
    return 0;
}
